src/webgpu/api/validation/render_pipeline/depth_stencil_state.spec.ts

```diff
@@ -50,13 +54,44 @@ g.test('depthCompare_optional')
     t.selectDeviceOrSkipTestCase(info.feature);
   })
   .fn(t => {
-    const { isAsync, format, depthCompare } = t.params;
+    const {
+      isAsync,
+      format,
+      depthCompare,
+      depthWriteEnabled,
+      stencilFrontDepthFailOp,
+      stencilBackDepthFailOp,
+    } = t.params;
     const info = kTextureFormatInfo[format];
     const descriptor = t.getDescriptor({
-      depthStencil: { format, depthCompare, depthWriteEnabled: false },
+      depthStencil: {
+        format,
+        depthCompare,
+        depthWriteEnabled,
+        stencilFront: { depthFailOp: stencilFrontDepthFailOp },
+        stencilBack: { depthFailOp: stencilBackDepthFailOp },
+      },
     });
 
-    t.doCreateRenderPipelineTest(isAsync, !(info.depth && depthCompare === undefined), descriptor);
+    const areDepthFailOpKeep =
+      stencilFrontDepthFailOp === 'keep' && stencilBackDepthFailOp === 'keep';
+
+    let success = false;
+    if (!info.depth) {
+      // depthCompare is optional for stencil-only formats.
+      if (!depthWriteEnabled) success = true;
+    } else {
+      // depthCompare is optional for formats with a depth if it's not used for anything.
+      if (depthWriteEnabled === false && areDepthFailOpKeep) success = true;
+      if (depthCompare) {
+        // validation will succeed normally for formats with depth aspect.
+        if (depthWriteEnabled && areDepthFailOpKeep) success = true;
+        // validation will succeed as well for formats with stencil and depth aspect.
+        if (info.stencil && depthWriteEnabled !== undefined) success = true;
+      }
+    }
```

Comment:
Apologies for another big change. It is much easier to think about written this way, but I still find it quite hard to pick apart where the different rules from the spec appear here. (I probably shouldn't have suggested inverting the usual pattern, as I realize there's a good reason we usually have `let success = true; if (...) success = false`.)

I think the most readable form would stick close to [the spec](https://raw.githack.com/gpuweb/gpuweb-pr-previews/b3df567de1e8e545a2c645a237d5c113be9d085f/gpuweb/gpuweb/4336/index.html#abstract-opdef-validating-gpudepthstencilstate):

```suggestion
    const depthFailOpsAreKeep =
      stencilFrontDepthFailOp === 'keep' && stencilBackDepthFailOp === 'keep';
    const stencilStateIsDefault = depthFailOpsAreKeep;

    let success = true;
    if (depthWriteEnabled || depthCompare !== 'always') {
      if (!info.depth) success = false;
    }
    if (!stencilStateIsDefault) {
      if (!info.stencil) success = false;
    }
    if (info.depth) {
      if (depthWriteEnabled === undefined) success = false;
      if (depthWriteEnabled || !depthFailOpsAreKeep) {
        if (depthCompare === undefined) success = false;
      }
    }
```

I would need you to test this to make sure it's right, but it follows the spec very closely so I hope it is.

Reply:
@kainino0x Thanks for the detailed answer. I followed your last suggestion with one nit:

```diff
-    if (depthWriteEnabled || depthCompare !== 'always') {
+    if (depthWriteEnabled || (depthCompare && depthCompare !== 'always')) {
```

I've updating spec to make it clear as well. See https://github.com/gpuweb/gpuweb/pull/4345

Reply:
@kainino0x (gentle ping)